### PR TITLE
Feature: firehose plugin support

### PR DESCRIPTION
**Pull request by peter-svensson**

### CODE REVIEW

**Comment by wenchajun** [comment]
Thanks for your contribution, but I didn't find firehose output plug-in in the official document, could you please provide me with a document or link?

**Comment by peter-svensson** [comment]
> Thanks for your contribution, but I didn't find firehose output plug-in in the official document, could you please provide me with a document or link?

Sure, it's part of the [AWS for fluentbit](https://github.com/aws/aws-for-fluent-bit), specifically [kinesis-firehose](https://github.com/aws/amazon-kinesis-firehose-for-fluent-bit).
So it's probably not an official plugin in that sense...

File: manifests/setup/setup.yaml [hunk 1]
```diff
@@ -4967,7 +5045,7 @@ spec:
         app.kubernetes.io/name: fluentbit-operator
     spec:
       containers:
-      - image: kubesphere/fluentbit-operator:latest
+      - image: kubesphere/fluentbit-operator:local
```
Comment (wanjunlei) [inline manifests/setup/setup.yaml:5048]: Why do need to change the tag?

File: manifests/setup/fluentbit-operator-deployment.yaml [hunk 2]
```diff
@@ -39,7 +39,7 @@ spec:
           mountPath: /var/run/docker.sock
       containers:
       - name: fluentbit-operator
-        image: 'kubesphere/fluentbit-operator:latest'
+        image: 'kubesphere/fluentbit-operator:local'
```
Comment (wanjunlei) [inline manifests/setup/fluentbit-operator-deployment.yaml:42]: Why do need to change the tag?

File: manifests/logging-stack/fluentbit-fluentBit.yaml [hunk 3]
```diff
@@ -6,7 +6,7 @@ metadata:
   labels:
     app.kubernetes.io/name: fluent-bit
 spec:
-  image: kubesphere/fluent-bit:v1.8.3
+  image: peter-fluentbit:v1.8.10
```
Comment (wanjunlei) [inline manifests/logging-stack/fluentbit-fluentBit.yaml:9]: V1.8.3 does not support this plugin, or there is a bug in v1.8.3? If true, we should build a fluent bit image of v1.8.10.

File: api/fluentbitoperator/v1alpha2/plugins/output/firehose_types.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+	"kubesphere.io/fluentbit-operator/api/fluentbitoperator/v1alpha2/plugins"
+	"kubesphere.io/fluentbit-operator/api/fluentbitoperator/v1alpha2/plugins/params"
+)
+
+// +kubebuilder:object:generate:=true
+
+// The Firehose output plugin, allows to ingest your records into AWS Firehose.
+// https://github.com/aws/amazon-kinesis-firehose-for-fluent-bit
+type Firehose struct {
```
Comment (wanjunlei) [inline api/fluentbitoperator/v1alpha2/plugins/output/firehose_types.go:14]: Variable naming needs to be adjusted. First, it is not recommended to use `_ `in the definition of crd, for example, `delivery_stream` should be changed to `deliveryStream`. Secondly, for acronyms, use uppercase, such as `Arn => ARN`, `Sts => STS`.

**Review by wanjunlei — CHANGES_REQUESTED** [review]

File: manifests/setup/fluentbit-operator-deployment.yaml (hunk 2 above)
Comment (peter-svensson) [reply manifests/setup/fluentbit-operator-deployment.yaml:42]: My bad, was for local testing 😊

File: manifests/setup/setup.yaml (hunk 1 above)
Comment (peter-svensson) [reply manifests/setup/setup.yaml:5048]: My bad, was for local testing 😊

File: api/fluentbitoperator/v1alpha2/plugins/output/firehose_types.go (hunk 4 above)
Comment (peter-svensson) [reply api/fluentbitoperator/v1alpha2/plugins/output/firehose_types.go:14]: Great - I'll fix!

File: manifests/logging-stack/fluentbit-fluentBit.yaml (hunk 3 above)
Comment (peter-svensson) [reply manifests/logging-stack/fluentbit-fluentBit.yaml:9]: Again, local testing.
I'm not sure there is an official docker image with the necessary plugins installed. 
One solution would be to build and distribute one that includes all the plugins that the operator supports, or just documenting how to build and use your own?
I think I prefer the first solution, but that requires a new version to be built and released for kubesphere/fluent-bit (not a big issue, I can create a PR for that in the correct repo as well of course).

As a side-note, I see that the `kubesphere` reops are archived, should we perhaps use `fluent/fluent-bit` instead in upcoming releases? Changing the names of the CRDs is trickier of course..

**Comment by peter-svensson** [comment]
Updated a bit, seems like I confused the old and new version of the firehose plugin.
The `kinesis_firehose` plugin seems to be supported in the `kubesphere/fluent-bit:v1.8.3` docker image so should work just fine.
Sorry for the confusion 😊

File: api/fluentbitoperator/v1alpha2/plugins/output/firehose_types.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	TimeKeyFormat *string `json:"timeKeyFormat,omitempty"`
+	// By default, the whole log record will be sent to Kinesis. If you specify a key name(s) with this option, then only those keys and values will be sent to Kinesis. For example, if you are using the Fluentd Docker log driver, you can specify data_keys log and only the log message will be sent to Kinesis. If you specify multiple keys, they should be comma delimited.
+	DataKeys *string `json:"dataKeys,omitempty"`
+	// By default, the whole log record will be sent to Firehose. If you specify a key name with this option, then only the value of that key will be sent to Firehose. For example, if you are using the Fluentd Docker log driver, you can specify log_key log and only the log message will be sent to Firehose.
+	LogKey *string `json:"logKey,omitempty"`
+	// ARN of an IAM role to assume (for cross account access).
+	RoleARN *string `json:"roleArn,omitempty"`
```
Comment (wanjunlei) [inline api/fluentbitoperator/v1alpha2/plugins/output/firehose_types.go:33]: `roleArn` should be `roleARN`

**Review by wanjunlei — CHANGES_REQUESTED** [review]

**Comment by wanjunlei** [comment]
By the way, please add the `sign-off for the commit.

**Review by wanjunlei — APPROVED** [review]

**Pull request merged**